### PR TITLE
Added error messages to asserts in std.range (part 2)

### PR DESCRIPTION
**Pull request by JackStouffer**

### CODE REVIEW

File: std/range/package.d [hunk 1]
```diff
@@ -7172,25 +7172,25 @@ private struct OnlyResult(T, size_t arity)
 
     T front() @property
     {
-        assert(!empty);
+        assert(!empty, "Attempting to fetch the front of an empty only");
```
Comment (quickfur) [inline std/range/package.d]: The unfortunate naming of `only` makes this sentence look really weird and ungrammatical (even though it's actually correct). Don't know how to fix this... perhaps put `only` in quotes?

Comment (schveiguy) [inline std/range/package.d]: We could switch to the convention: "<rangename>.front -- range is empty", e.g. "only.front -- range is empty"

**Comment by JackStouffer** [comment]
@quickfur clarified

File: std/range/package.d [hunk 2]
```diff
@@ -7224,7 +7224,14 @@ private struct OnlyResult(T, size_t arity)
         OnlyResult result = this;
         result.frontIndex += from;
         result.backIndex = this.frontIndex + to;
-        assert(from <= to && to <= length);
+        assert(
+            from <= to,
+            "Attempting to slice an only with a larger first argument than the second."
```
Comment (quickfur) [inline std/range/package.d]: What about "Only range" here too?

File: std/range/package.d [hunk 3]
```diff
@@ -7224,7 +7224,14 @@ private struct OnlyResult(T, size_t arity)
...
+        );
+        assert(
+            to <= length,
+            "Attempting to slice using an out of bounds index on an only"
```
Comment (quickfur) [inline std/range/package.d]: and here?

File: std/range/package.d [hunk 4]
```diff
@@ -7264,7 +7287,7 @@ private struct OnlyResult(T, size_t arity : 1)
 
     T opIndex(size_t i)
     {
-        assert(!_empty && i == 0);
+        assert(!_empty && i == 0, "Attempting to fetch an out of bounds index from an only");
```
Comment (quickfur) [inline std/range/package.d]: ditto

File: std/range/package.d [hunk 5]
```diff
@@ -7275,7 +7298,14 @@ private struct OnlyResult(T, size_t arity : 1)
 
     OnlyResult opSlice(size_t from, size_t to)
     {
-        assert(from <= to && to <= length);
+        assert(
+            from <= to,
+            "Attempting to slice an only with a larger first argument than the second."
```
Comment (quickfur) [inline std/range/package.d]: ditto

File: std/range/package.d [hunk 6]
```diff
@@ -7275,7 +7298,14 @@ private struct OnlyResult(T, size_t arity : 1)
...
+        );
+        assert(
+            to <= length,
+            "Attempting to slice using an out of bounds index on an only"
```
Comment (quickfur) [inline std/range/package.d]: ditto

File: std/range/package.d (hunk 1 above)
Comment (quickfur) [inline std/range/package.d]: yeah, this might be good for better uniformity across error messages from the different ranges.

**Comment by JackStouffer** [comment]
@quickfur Fixed

**Comment by quickfur** [comment]
LGTM. Thanks!

**Comment by quickfur** [comment]
Auto-merge toggled on

**Comment by WalterBright** [comment]
Auto-merge toggled off

**Comment by WalterBright** [comment]
I'm not much in favor of this. They had a heapin' helpin' of bloat to the executable, especially if this is carried all across Phobos. They're also adding trivial explanations for things that are trivially obvious, much as:
```
++i; // increment i
```
type comments are not good. I appreciate what you're trying to do, and the work you did on improving the imports was very good. May I suggest instead looking at functions in Phobos modules and ensure each has a `Params:` and `Returns:` ddoc comment, and perhaps even ensure they each have an illuminating example?

**Comment by WalterBright** [comment]
May I also ask that you not use PR names like `patch-crazydiamond` that give no hint what the PR is about? I ask because I look at PRs often via https://auto-tester.puremagic.com/pulls.ghtml?projectid=1

Having a name that gives a hint would be welcome. Thanks!

**Comment by quickfur** [comment]
@WalterBright  I'm not sure I agree with that sentiment. After all, these messages are in asserts, and (presumably?) the compiler would leave them out in release builds. So the bloat problem isn't too big of a deal IMO.  Besides, asserts with no messages is very unfriendly in a library, because the user can't be expected to understand library internals enough to be able to ascertain what a failed assertion in line 1234 of some obscure Phobos implementation code might signify.

**Comment by JackStouffer** [comment]
My problem is that a are amount of the errors users get are completely unhelpful.

For example

```
void main()
{
    import std.range : only;
    auto a = only('a');
    a.popFront();
    a.front;
}
```
Yeilds this on macOS

```
core.exception.AssertError@std/range/package.d(6860): Assertion failure
----------------
4   dmd_runGEA74i                       0x000000010b84aeb4 _d_assert + 104
5   dmd_runGEA74i                       0x000000010b8977aa void std.range.__assert(int) + 38
6   dmd_runGEA74i                       0x000000010b841f95 pure nothrow @property @nogc @safe char std.range.OnlyResult!(char, 1uL).OnlyResult.front() + 93
7   dmd_runGEA74i                       0x000000010b841e69 _Dmain + 37
8   dmd_runGEA74i                       0x000000010b85dfff D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv + 39
9   dmd_runGEA74i                       0x000000010b85df33 void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) + 35
10  dmd_runGEA74i                       0x000000010b85dfa4 void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() + 44
11  dmd_runGEA74i                       0x000000010b85df33 void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) + 35
12  dmd_runGEA74i                       0x000000010b85de99 _d_run_main + 497
13  dmd_runGEA74i                       0x000000010b841f03 main + 15
14  libdyld.dylib                       0x00007fff965065ac start + 0
15  ???                                 0x0000000000000000 0x0 + 0
```

This is a contrived example, as I have only one `only` in my code and it's four lines long, But I have small projects (~2 KLOC) which heavily use ranges, and these sorts of errors pop up all of the time. In general I find the D experience with errors and error reporting to be sub-par. Asserts and enforce calls without error messages are common in Phobos code.

I really, really don't want to have to open my local copy of Phobos, checkout the last release, and go to the line in order to figure out whats going wrong in my program. That's unacceptable.

**Comment by JackStouffer** [comment]
For context, here's what the user sees with this change

```
core.exception.AssertError@dmd/src/../../phobos/std/range/package.d(7259): Attempting to fetch the front of an empty Only range
----------------
4   dmd_runAkJFuX                       0x00000001041e5cff onAssertErrorMsg + 107
5   dmd_runAkJFuX                       0x00000001041e5f50 _d_assert_msg + 68
6   dmd_runAkJFuX                       0x00000001041d7869 pure nothrow @property @nogc @safe char std.range.OnlyResult!(char, 1uL).OnlyResult.front() + 133
7   dmd_runAkJFuX                       0x00000001041d7605 _Dmain + 37
8   dmd_runAkJFuX                       0x00000001041fdb23 D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv + 39
9   dmd_runAkJFuX                       0x00000001041fda2f void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) + 51
10  dmd_runAkJFuX                       0x00000001041fdac1 void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() + 53
11  dmd_runAkJFuX                       0x00000001041fda2f void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) + 51
12  dmd_runAkJFuX                       0x00000001041fd99b _d_run_main + 831
13  dmd_runAkJFuX                       0x00000001041d76c1 main + 33
14  libdyld.dylib                       0x00007fff965065ac start + 0
15  ???                                 0x0000000000000000 0x0 + 0
```

Much better!

I still would like line numbers though :)

**Comment by schveiguy** [comment]
> I really, really don't want to have to open my local copy of Phobos, checkout the last release, and go to the line in order to figure out whats going wrong in my program

Hm... I just had a thought. What if an error that's thrown in phobos/druntime, instead of showing the error as "std/range/package.d:...", showed it as "https://github.com/dlang/phobos/blob/2c4e4c76ce0f5ce91581c6a42f9cb2a1c9d53105/std/range/package.d:..."?

Then you can copy and paste the error into a browser and see the exact file that has caused the error. Such a mechanism would only be employed if the compiler is told to do so (i.e. we'd only do this for official releases).

**Comment by quickfur** [comment]
Wouldn't this introduce even *more* bloat into the executable, now that we need to also store git hashes in assert messages?

**Comment by JackStouffer** [comment]
> Then you can copy and paste the error into a browser

NO!

We cannot ask users to open a browser to decipher an error message, and then ask them to read the Phobos source code to understand why their program blew up. That's ridiculous.

We should just have good error messages.

Sorry if I seem a little more animated than usual, it's just that the debugging experience for Mac users sucks, and anything that helps me spend less time on manual debugging is a huge win for me.

**Comment by wilzbach** [comment]
> Wouldn't this introduce even more bloat into the executable, now that we need to also store git hashes in assert messages?

I thought @schveiguy suggested a simple replacement of <file> to github.com/dlang/phobos/blob/<hash>/... 
Imho that's a great idea to improve the usefulness of error messages, but I agree with @JackStouffer that if the error messages would be better we don't need the magic in the first place. Don't forget that difficult to understand error messages are among the top 3 points that are mentioned as reasons against D.

**Comment by schveiguy** [comment]
> Wouldn't this introduce even more bloat into the executable,

Why? You only have to store the prefix for the url once (per library, so one for phobos one for druntime).

> We cannot ask users to open a browser to decipher an error message, and then ask them to read the Phobos source code to understand why their program blew up. That's ridiculous.

It's not that bad actually. If you can find the code, you can see why the code is failing. Even your error messages may not be enough "yeah, I get that I tried to access front of this empty range, but why the hell is it empty?". Debugging isn't always easy.

> the debugging experience for Mac users sucks

Totally with you on that.

**Comment by WalterBright** [comment]
Yields this on Windows:
```
core.exception.AssertError@std\range\package.d(7250): Assertion failure
----------------
0x00403C27 in _d_assert
0x0040202F in _Dmain at C:\cbx\mars\bug\test20.d(6)      <==== note the line number !!!!
0x004026D3 in D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv
0x00402697 in void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll()
0x00402598 in _d_run_main
0x00402388 in main at C:\cbx\mars\bug\test20.d(7)
0x00419B05 in mainCRTStartup
0x76AD338A in BaseThreadInitThunk
0x77239902 in RtlInitializeExceptionChain
0x772398D5 in RtlInitializeExceptionChain
```
You get source files and line numbers for the modules compiled with `-g`.

> I really, really don't want to have to open my local copy of Phobos, checkout the last release, and go to the line in order to figure out whats going wrong in my program.

It would be strange to have to go to all that trouble, because in order to run `dmd` the correct source code has to already be there, as the compiler reads it! For me, it would be:
```
me ..\phobos\src\std\range\package.d
Ctrl-X Ctrl-L 7250  (that's my editor command to go to line 7250)
```

**Comment by JackStouffer** [comment]
> You get source files and line numbers for the modules compiled with -g.

```
$ dmd -g -run test.d
core.exception.AssertError@std/range/package.d(6860): Assertion failure
----------------
4   dmd_runnJnbtG                       0x000000010b17ffb0 _d_assert + 104
5   dmd_runnJnbtG                       0x000000010b1c6662 void std.range.__assert(int) + 38
6   dmd_runnJnbtG                       0x000000010b167029 pure nothrow @property @nogc @safe char std.range.OnlyResult!(char, 1uL).OnlyResult.front() + 93
7   dmd_runnJnbtG                       0x000000010b166dd8 _Dmain + 40
8   dmd_runnJnbtG                       0x000000010b193127 D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv + 39
9   dmd_runnJnbtG                       0x000000010b19305b void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) + 35
10  dmd_runnJnbtG                       0x000000010b1930cc void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() + 44
11  dmd_runnJnbtG                       0x000000010b19305b void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) + 35
12  dmd_runnJnbtG                       0x000000010b192fc1 _d_run_main + 497
13  dmd_runnJnbtG                       0x000000010b166f99 main + 33
14  libdyld.dylib                       0x00007fff965065ac start + 0
15  ???                                 0x0000000000000000 0x0 + 0
```

Isn't macOS great?

> It would be strange to have to go to all that trouble, because in order to run dmd the correct source code has to already be there, as the compiler reads it!

Ok, I admit that I forgot about the local copy DMD uses.

Still I really don't want to read the source of the standard library in order to figure out what's wrong. No user friendly language does this.

**Comment by quickfur** [comment]
@schveiguy You're right, I concede.

@WalterBright Line numbers or not, it's still very unfriendly to have a user's program abort with a line number *inside library code not written by the user*.   I don't have a problem with just a filename and line number when a problem occurs in the user's code, but we're talking about the standard library here. I think the standard for the standard(!) library ought to be higher. The standard library should do its best to make the user's life less miserable by pointing out the reason for failures, especially in frequently-used library code.

Besides, the whole bloat argument is moot anyway, seeing that we still haven't resolved the gigantic template symbol problem, just *one* instance of which probably already far overwhelms the amount of "bloat" this PR would add to Phobos.

**Comment by JackStouffer** [comment]
> Besides, the whole bloat argument is moot anyway, seeing that we still haven't resolved the gigantic template symbol problem, just one instance of which probably already far overwhelms the amount of "bloat" this PR would add to Phobos.

Solving that and making DMD fully lazy would probably cut the average binary size in half

**Comment by quickfur** [comment]
To make the problem more obvious, consider an UFCS chain spanning several lines:
````
auto r = myDataSource.getRange()
        .filter!(e => someComplexCondition(e))
        .map!(e => e.selectedAttribute)
        .take(endIdx - startIdx)
        .chain(prevDataSet)
        .map!(e => outputFmt.format(e))
        .joiner(delimiter);
````
AFAIK, if an assert fails in *any one of* the components of the pipeline, the same line number will show up for this statement in the user's code. That is not much help in narrowing down the problem.  Now, you may argue that the failing line number in Phobos should indicate which component failed, but again, it's unreasonable to expect the user to have to read (and understand) the standard library's code just so he can figure out what we, the Phobos maintainers, have already figured out and can simply convey by adding a helpful error message to the `assert`.

**Comment by ZombineDev** [comment]
@JackStouffer About the binary bloat - no, the biggest problem is that name mangling produces symbols with exponential size. Lazy compilation is orthogonal to that (though still an important improvement) because the linker can delete unused symbols, but cannot change used ones.
Rainer's awesome [PR](https://issues.dlang.org/show_bug.cgi?id=15831#c8) obliterates symbol bloat by turning a 250 MB binary into a 645 KB one! Note that the same executable was 350-400MB on Linux, so there's really nothing close to symbols w.r.t bloating the binary size.

So FWIW, I don't think we should worry that putting helpful error messages in `assert`s would the increase the binary size.

**Comment by WalterBright** [comment]
1. It is true that `-release` will remove the bloat along with the `assert`. But many users wish to leave the asserts in their released code, for good reason. I seriously doubt they wish to have the bloat as well.

2.  `No user friendly language does this` is a bit of a tautology. Nevertheless, most libraries have opaque interfaces and return error codes. Having a file and line number means it is only two steps away to bringing on screen, where the `assert` will presumably be documented and there'll be surrounding context. I tend to view Phobos modules as more intimately connected with the user's code than usual programming language libraries. Despite gcc being open source, finding the right bit of source corresponding to a library call can be quite a chore (I've tried). In D, it's trivial.

3. If you're running under a debugger with symbol debug info on, it will drop you right there in the library source.

4. I agree that the missing line numbers on OSX is an unfortunate bug (is it in bugzilla?), but this is not the right fix for that.

5. Saying that since there are other sources of bloat, that this bloat is insignificant, is an argument I have strong objection to. Phobos/dmd are complex enough that you can find a justification for any practice whatsoever.

6. Certainly there are benefits to these messages. But there are downsides, and in a systems programming language that people need to be lean and mean at runtime, it's a poor tradeoff.

**Comment by JackStouffer** [comment]
>  Nevertheless, most libraries have opaque interfaces and return error codes.

Having error codes, while still not a good idea, would be better than what we have now. Error codes can typically be googled or looked up in some physical or electronic reference. The code is typically accompanied by the description of what went wrong.

D should not require the user to read someone else's code in order to understand their error.

> Having a file and line number means it is only two steps away to bringing on screen, where the assert will presumably be documented and there'll be surrounding context.

No other programming language that I have used requires opening the source of the standard library.

If I tried pitching D to my co-workers, and they realized that in order to debug the library's sanity checks, they needed to pull open the source, I would be laughed out of the building.

> If you're running under a debugger with symbol debug info on, it will drop you right there in the library source.

Debuggers don't work on macOS: https://issues.dlang.org/show_bug.cgi?id=14927

That just says GDB but LLDB doesn't work either.

> I agree that the missing line numbers on OSX is an unfortunate bug (is it in bugzilla?), but this is not the right fix for that.

https://issues.dlang.org/show_bug.cgi?id=8172

**Comment by JackStouffer** [comment]
> If you're running under a debugger with symbol debug info on, it will drop you right there in the library source.

To elaborate on this, look at this from a beginner's perspective:

D was the first systems language I have used. I starting learning it a little over a year ago specifically to learn more about the lower level operations of computers. At that point, I have never heard of GDB or LLDB. If I'm starting out, and I get an error like the one above, I would literally not know what to do other than to head over to stack overflow, and hopefully someone like Adam would be there to help.

That's assuming I would have the drive to go to SO rather than just give up and use something else, or go back to Python.

Assume I had the wherewithal to know to go look at the source. If I got to this line https://github.com/dlang/phobos/pull/4573/files#diff-43f17b39f02da9efd68008101ca4cd81R7227, it's totally possible I don't know what `opSlice` is!

**Comment by JackStouffer** [comment]
One more point to consider: if the DMD front-end was made lazy, these error messages would not be included in the binary unless the function was used.

**Comment by WalterBright** [comment]
> Error codes can typically be googled or looked up in some physical or electronic reference. The code is typically accompanied by the description of what went wrong in plain English.

Having done much of that in my career, I am surprised that you'd find it easier than opening the specified file and going to a line number. I don't really understand the expressed fear of looking at source code. Phobos source code is supposed to be educational. Looking at others' source code is how one learns. Besides, at that line should be a comment explaining the issue in plain English. Everything else can be ignored.

> if the DMD front-end was made lazy, these error messages would not be included in the binary unless the function was used.

This is already how linkable libraries work.

> No other programming language that I have used requires opening the source of the standard library.

Opaque libraries are the norm in other languages. But Phobos is not meant to be opaque.

**Comment by andralex** [comment]
I'll allow it. It's the civilized thing to do. The intent was to add messages to all asserts, but when I wrote the code I was always in a hurry.

    git grep assert | wc -l

returns 25K instances. For an average of 40 bytes per message, we're looking at 2M of messages - which is nothing for a debug build today.

I'd like to thank @JackStouffer and @atilaneves for their great contributions as of late.
